scheds/rust/scx_bpfland/src/main.rs

```diff
@@ -0,0 +1,265 @@
+use std::sync::atomic::Ordering;
+use std::sync::Arc;
+use std::time::Duration;
+
+use std::str;
+
+use anyhow::Context;
+use anyhow::Result;
+use clap::Parser;
+use log::info;
+
+use metrics::{gauge, Gauge};
+use metrics_exporter_prometheus::PrometheusBuilder;
+
+use rlimit::{getrlimit, setrlimit, Resource};
+
+use libbpf_rs::skel::OpenSkel;
+use libbpf_rs::skel::Skel;
+use libbpf_rs::skel::SkelBuilder;
+
+use scx_utils::scx_ops_attach;
+use scx_utils::scx_ops_load;
+use scx_utils::scx_ops_open;
+use scx_utils::uei_exited;
+use scx_utils::uei_report;
+use scx_utils::UserExitInfo;
+
+const SCHEDULER_NAME: &'static str = "scx_bpfland";
+
+const VERSION: &'static str = env!("CARGO_PKG_VERSION");
+
+/// scx_bpfland: a vruntime-based sched_ext scheduler that prioritizes interactive workloads.
+///
+/// This scheduler is derived from scx_rustland, but it is fully implemented in BFP with minimal
+/// user-space part written in Rust to process command line options, collect metrics and logs out
+/// scheduling statistics.
+///
+/// The BPF part makes all the scheduling decisions (see src/bpf/main.bpf.c).
+#[derive(Debug, Parser)]
+struct Opts {
+    /// Exit debug dump buffer length. 0 indicates default.
+    #[clap(long, default_value = "0")]
+    exit_dump_len: u32,
+
+    /// Maximum scheduling slice duration in microseconds.
+    #[clap(short = 's', long, default_value = "20000")]
+    slice_us: u64,
+
+    /// Minimum scheduling slice duration in microseconds.
+    #[clap(short = 'S', long, default_value = "500")]
+    slice_us_min: u64,
+
+    /// Threshold of voluntary context switch per second, used to classify interactive tasks
+    /// (0 = disable interactive tasks classification).
+    #[clap(short = 'c', long, default_value = "10")]
+    nvcsw_thresh: u64,
+
+    /// Priority boost for interactive tasks (0 = disable priority boost).
+    #[clap(short = 'b', long, default_value = "100")]
+    weight_boost: u64,
+
+    /// Disable direct dispatch via built-in idle selection logic (can provide better performance
+    /// on non-SMT architectures).
+    #[clap(short = 'i', long, action = clap::ArgAction::SetFalse)]
+    builtin_idle: bool,
+
+    /// Enable the Prometheus endpoint for metrics on port 9000.
+    #[clap(short = 'p', long, action = clap::ArgAction::SetTrue)]
+    enable_prometheus: bool,
+
+    /// Enable BPF debugging via /sys/kernel/debug/tracing/trace_pipe.
+    #[clap(short = 'd', long, action = clap::ArgAction::SetTrue)]
+    debug: bool,
+
+    /// Enable verbose output, including libbpf details.
+    #[clap(short = 'v', long, action = clap::ArgAction::SetTrue)]
+    verbose: bool,
+
+    /// Print scheduler version and exit.
+    #[clap(short = 'V', long, action = clap::ArgAction::SetTrue)]
+    version: bool,
+}
+
+struct Metrics {
+    nr_running: Gauge,
+    nr_kthread_dispatches: Gauge,
+    nr_direct_dispatches: Gauge,
+    nr_prio_dispatches: Gauge,
+    nr_shared_dispatches: Gauge,
+}
+
+impl Metrics {
```

Comment:
Nice to see that you are using the metrics framework!

Reply:
@jfernandez yep, thanks for adding it! I'm planning to use the metrics in scx_rustland as well.